File: src/main/java/me/zeroeightsix/kami/util/HueCycler.java
package me.zeroeightsix.kami.util;

import org.lwjgl.opengl.GL11;

import java.awt.*;

/**
 * Created by 086 on 22/01/2018.
 */
public class HueCycler {

    int index = 0;
    int[] cycles;

    public HueCycler(int cycles) {
        if (cycles <= 0) throw new IllegalArgumentException("cycles <= 0");
        this.cycles = new int[cycles];
        double hue = 0;
        double add = 1 / (double) cycles;
        for (int i = 0; i < cycles; i++) {
            this.cycles[i] = Color.HSBtoRGB((float) hue, 1, 1);
            hue += add;
        }
    }

    public void reset() {
        index = 0;
    }

    public void reset(int index) {
        this.index = index;
    }

    public int next() {
        int a = cycles[index];
        index++;
        if (index >= cycles.length) index = 0;
        return a;
    }

    public void setNext() {
        int rgb = next();
    }

    public void set() {
        int rgb = cycles[index];
        float red = ((rgb >> 16) & 0xFF) / 255f;
        float green = ((rgb >> 8) & 0xFF) / 255f;
        float blue = (rgb & 0xFF) / 255f;
        GL11.glColor3f(red, green, blue);
    }

    public void setNext(float alpha) {
        int rgb = next();
        float red = ((rgb >> 16) & 0xFF) / 255f;
        float green = ((rgb >> 8) & 0xFF) / 255f;
        float blue = (rgb & 0xFF) / 255f;
        GL11.glColor4f(red, green, blue, alpha);
    }

    public int current() {
        return cycles[index];
    }
}

File: src/main/kotlin/org/kamiblue/client/util/color/HueCycler.kt
package org.kamiblue.client.util.color

import org.kamiblue.client.util.color.ColorConverter.hexToRgb
import org.lwjgl.opengl.GL11
import java.awt.Color

class HueCycler(val cycles: Int) {
    private val hueMultiplier = 1f / cycles.toFloat()
    private val colorCycle: Array<Int> = Array(cycles) { i ->
        Color.HSBtoRGB(i * hueMultiplier, 1f, 1f)
    }
    private var index = 0

    fun reset() {
        set(0)
    }

    fun set(indexIn: Int) {
        index = indexIn
    }

    fun setCurrent() {
        val rgb = colorCycle[index]
        val red = (rgb shr 16 and 0xFF) / 255f
        val green = (rgb shr 8 and 0xFF) / 255f
        val blue = (rgb and 0xFF) / 255f
        GL11.glColor3f(red, green, blue)
    }

    fun setNext() {
        setNext(1f)
    }

    fun setNext(a: Float) {
        inc()
        val rgb = currentRgb()
        val r = (rgb.r shr 16 and 0xFF) / 255f
        val g = (rgb.g shr 8 and 0xFF) / 255f
        val b = (rgb.b and 0xFF) / 255f
        GL11.glColor4f(r, g, b, a)
    }

    fun currentInt(): Int {
        return colorCycle[index]
    }

    fun currentRgba(alpha: Int): ColorHolder {
        val color = currentRgb()
        color.a = alpha
        return color
    }

    fun currentRgb(): ColorHolder {
        return hexToRgb(currentInt())
    }

    operator fun plus(plus: Int) {
        index += plus
        if (index >= cycles) index = 0
    }

    operator fun inc(): HueCycler {
        index++
        if (index >= cycles) index = 0
        return this
    }

    init {
        require(cycles > 0) { "cycles <= 0" }
    }
}